app/services/oauth2_refresher_service.rb

```diff
@@ -13,8 +15,13 @@ def self.build
     new
   end
 
-  sig { override.params(connection: UpholdConnection).returns(T.any(Oauth2RefresherService::Success, BFailure)) }
+  sig { override.params(connection: UpholdConnection).returns(Oauth2RefresherService::TYPES) }
   def call(connection)
+    if connection.refresh_token.nil?
+      connection.record_refresh_failure!
+      return BFailure.new(errors: ["Cannot refresh without refresh token"])
+    end
+
```

Comment:
@tsmartt That's not what it's doing.  It is intended to handle all oauth2 client credentials spec methods eventually, but  right now it is only handling refresh tokens.  The intended direction of `ClientCredentials` is to have an `access_token` and `code` method alongside the `refresh_token` because that is the logical grouping of possible grant types that is associated with the client credentials. See the spec for the client credentials grant: https://datatracker.ietf.org/doc/html/rfc6749#section-4.4

If we encounter a wallet authentication strategy that is not using ClientCredentials I will write another class that handles that specific flow and update the various jobs accordingly and probably subclass the basic HTTP request mechanisms to be shared between the different OAuth2 flows.  Broadly, I am solving for all cases of client credentials grant flows so we don't have to keep reimplementing the same logic in different classes.

Reply:
Oauth2::ClientCredentials is an HTTPClient, not a service.  I write them very differently because their concerns are quite different.